:seedling: Move Kubeadm API v1beta1 webhooks to separate package

odvarkadaniel opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/guide.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:
To work towards #9011, we need to remove the reliance on controller-runtime from the API packages.
This updates the webhooks to use the CustomDefaulter and CustomValidator pattern and moves them to match the Cluster and ClusterClass webhooks. This Is a more major lift but removes quite a bit of the code from the API package itself, making it's reliance on controller-runtime a lot smaller.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->

odvarkadaniel commented:
/area api

JoelSpeed reviewed (commented):
Changes look as expected, if we can work out the verify failures then this should be good to go, thanks for picking this up @odvarkadaniel

sbueringer commented on internal/webhooks/kubeadm_control_plane.go:
```diff
@@ -14,9 +14,10 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
```
Please also already add import-restriction files to the api packages (see https://github.com/kubernetes-sigs/cluster-api/pull/9407)

odvarkadaniel replied:
Will do!

odvarkadaniel commented:
Any idea about the verify issues?

Command `make generate` seems to be generating a lot of unnecessary stuff.

killianmuldoon commented:
> Any idea about the verify issues?
Command make generate seems to be generating a lot of unnecessary stuff.

The generate command regenerates everything. Once you've done that and committed the result - which should be relatively small as the vast majority of files won't change - you should be able to pass the `verify` check.

odvarkadaniel commented:
Is the failure for `pull-cluster-api-e2e-main` ok in this case?

If so, then this should be ready for a review:
-  Moved control-plane/kubeadm webhooks
- Moved bootstrap/kubeadm webhooks
- Added the import restrictions for these

JoelSpeed commented on internal/webhooks/kubeadmcontrolplanetemplate.go:
I suspect rather than grouping this webhook into the top level webhooks folder here, it should be in a webhooks folder in the kubeadm folder, so that the pattern looks the same, so, `controlplane/kubeadm/internal/webhooks` for this one rather than in this folder, WDYT?

sbueringer replied:
+1

odvarkadaniel replied:
This is pretty challenging to do without running into import cycle or invalid internal import usage. I am currently almost done, with hopefully a sensible solution, but...

What I need to figure out is where to place functions used within `Validate()`, which I moved into `kubeadmconfig_types.go`. There is 3 of them (`validateFiles, validateUsers, validateIgnition`).

JoelSpeed reviewed (commented):
I think merging all of the webhooks into the top `internal/webhooks` folder isn't our intention, and instead, we should keep `internal/webhooks` folders into each of the subprojects within the repo as it were, can you see how much work it is to move them back and create that same pattern in the different subprojects?

sbueringer commented on config/webhook/manifests.yaml:
```diff
@@ -48,6 +48,91 @@ webhooks:
     resources:
     - clusterclasses
   sideEffects: None
+- admissionReviewVersions:
```
There should be at most changes in the order I think

odvarkadaniel replied:
I don't see any duplications in there. I just shuffled with the kubebuilder tags when necessary.

JoelSpeed commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -107,6 +108,17 @@ type KubeadmConfigSpec struct {
 	Ignition *IgnitionSpec `json:"ignition,omitempty"`
 }
 
+// Validate ensures the KubeadmConfigSpec is valid.
+func (c *KubeadmConfigSpec) Validate(pathPrefix *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, validateFiles(c, pathPrefix)...)
+	allErrs = append(allErrs, validateUsers(c, pathPrefix)...)
+	allErrs = append(allErrs, validateIgnition(c, pathPrefix)...)
+
+	return allErrs
+}
```
I wouldn't have expected a validate function to be a member of `KubeadmConfigSpec`, I don't recall doing this for others?

sbueringer replied:
I have to take a closer look but KubeadmConfigSpec is a very special case as we need the validation in KCP as well (IIRC). We also have to use the default func in KCP.

odvarkadaniel replied:
As @sbueringer is mentioning, we are using the Validate function all over the place.
I do think that in this case, the Validate function being a member of `KubeadmConfigSpec` is not a bad idea. On the other hand, we would also need to move `validate{Files, Users, Ignition}` to some other accessible file (not internal).

What I have right now locally is that I just copied this code from `bootstrap/kubeadm` to also `controlplane/kubeadm` which uses it. I did the same thing for the defaulting function in KCP. 

I will try to think about some more sensible way to do this. If you have any ideas, please let me know.

sbueringer replied:
I would be in favor of keeping the `DefaultKubeadmConfigSpec` and `(c *KubeadmConfigSpec) validate(name string)` funcs in the v1beta1 API package (and just move them from the _webhook to the _types file) (also includes the funcs that validate is calling)

I think this should allow us to move the webhooks without having to do complex refactorings

sbueringer replied:
The goal of this issue/PR is mainly to get rid of the controller-runtime dependencies, we can do that without touching these functions.

JoelSpeed replied:
Sure, if that drops the CR dep then lets do that for now and see how we get on

odvarkadaniel replied:
Wouldn't it be better to move the functions that Validate is calling to move somewhere else than the _types file, just to not pollute the file that much?

sbueringer replied:
Fine for me

JoelSpeed commented on bootstrap/kubeadm/controllers/alias.go:
Do we use a controllers alias for the top level webhooks or do we have a more generic alias? Wondering if there should be a webhooks alias and then webhooks internal as you've done

sbueringer replied:
Yup. These should be added in bootstrap/kubeadm/webhooks/alias.go

JoelSpeed commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfig.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (webhook *KubeadmConfig) validate(c bootstrapv1.KubeadmConfigSpec, name string) error {
+	allErrs := c.Validate(field.NewPath("spec"))
```
Any reason not to have the validation in this field rather than in the API? The issue with having it in the API is it then adds code to the API and that will cause dependencies. I think the validation should all be moved over here

sbueringer replied:
Let's finish the discussion above first. But I think we can move the ValidateCreate/Update/Delete funcs without touching this code

sbueringer replied:
I think in any case it makes sense to do one PR first where we only move the funcs of the webhook interfaces. That removes the dependencies on controller-runtime. We can then discuss if we want to move more

JoelSpeed commented on controlplane/kubeadm/internal/controllers/alias.go:
Should be a webhooks alias in a public not internal package no?

JoelSpeed commented on controlplane/kubeadm/internal/filters.go:
```diff
@@ -215,8 +216,8 @@ func matchInitOrJoinConfiguration(machineConfig *bootstrapv1.KubeadmConfig, kcp
 	// But that is something that is ensured by our API guarantees.
-	bootstrapv1.DefaultKubeadmConfigSpec(kcpConfig)
-	bootstrapv1.DefaultKubeadmConfigSpec(&machineConfig.Spec)
+	util.DefaultKubeadmConfigSpec(kcpConfig)
```
Is this the same package others have their defaulting in? Or do we normally construct the defaulting webhook and use that to default the spec?

JoelSpeed commented on util/util.go:
```diff
@@ -634,3 +635,16 @@ func MergeMap(maps ...map[string]string) map[string]string {
 	return m
 }
+
+// DefaultKubeadmConfigSpec defaults a KubeadmConfigSpec.
+func DefaultKubeadmConfigSpec(r *bootstrapv1.KubeadmConfigSpec) {
+	if r.Format == "" {
+		r.Format = bootstrapv1.CloudConfig
+	}
+	if r.InitConfiguration != nil && r.InitConfiguration.NodeRegistration.ImagePullPolicy == "" {
+		r.InitConfiguration.NodeRegistration.ImagePullPolicy = "IfNotPresent"
+	}
+	if r.JoinConfiguration != nil && r.JoinConfiguration.NodeRegistration.ImagePullPolicy == "" {
+		r.JoinConfiguration.NodeRegistration.ImagePullPolicy = "IfNotPresent"
+	}
+}
```
I think this just wants to be in the defaulting webhook itself, then use the webhook in other packages that need this defautling

odvarkadaniel commented:
I have hopefully addressed all the comments and this should be therefore ready to review.

sbueringer commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -107,6 +121,219 @@ type KubeadmConfigSpec struct {
+	return allErrs
+}
+
+func validateFiles(c KubeadmConfigSpec, pathPrefix *field.Path) field.ErrorList {
```
```suggestion
func (c *KubeadmConfigSpec) validateFiles(pathPrefix *field.Path) field.ErrorList {
```

I would suggest to keep these 3 funcs as methods. I don't see the benefit in passing in KubeadmConfigSpec

(maybe I'm missing something)

odvarkadaniel replied:
Yes, makes sense. I will change that.

sbueringer commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -107,6 +121,219 @@ type KubeadmConfigSpec struct {
+			)
+		}
+	}
+
```
Did we intentionally drop the following?
```

	if c.UseExperimentalRetryJoin {
		allErrs = append(
			allErrs,
			field.Forbidden(
				pathPrefix.Child("useExperimentalRetryJoin"),
				cannotUseWithIgnition,
			),
		)
	}

```

odvarkadaniel replied:
Yes, I dropped this part also with the test case as this flag is deprecated and the linter was screaming about this and the documentation for this part said, that this flag is no longer needed.

sbueringer replied:
Please keep it. This field is deprecated on the API but CAPI users can still use it until we drop it in v1beta2. Before that we should not drop the corresponding validation. Please check the golangci.yaml we probably have an exclude for it (just not for this package)

sbueringer commented on bootstrap/kubeadm/internal/webhooks/doc.go:
```diff
@@ -0,0 +1,18 @@
+limitations under the License.
+*/
+
+// Package webhooks provides the validating webhook for KubeadmConfig scale subresource.
```
```suggestion
// Package webhooks provides the validating webhooks for KubeadmConfig and KubeadmConfigTemplate.
```

sbueringer commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfig_test.go:
```diff
@@ -335,58 +340,44 @@ func TestKubeadmConfigValidate(t *testing.T) {
 			},
 			expectErr: true,
 		},
```
Same here, why did we drop this test?

odvarkadaniel replied:
This got dropped as the `UseExperimentalRetryJoin` is deprecated.

sbueringer commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfigtemplate.go:
```diff
@@ -14,52 +14,73 @@ See the License for the specific language governing permissions and
+func (webhook *KubeadmConfigTemplate) Default(_ context.Context, obj runtime.Object) error {
+	c, ok := obj.(*bootstrapv1.KubeadmConfigTemplate)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", obj))
```
```suggestion
		return apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfigTemplate but got a %T", obj))
```

sbueringer replied:
Same below

sbueringer commented on bootstrap/kubeadm/webhooks/doc.go:
```diff
@@ -0,0 +1,18 @@
+limitations under the License.
+*/
+
+// Package webhooks provides the validating webhook for KubeadmConfig scale subresource.
```
```suggestion
// Package webhooks provides the validating webhooks for KubeadmConfig and KubeadmConfigTemplate.
```

sbueringer commented on bootstrap/kubeadm/api/.import-restrictions:
```diff
@@ -0,0 +1,5 @@
+rules:
```
Let's also add an additional restrictions file into the v1beta package (the same as api/v1beta1/.import-restrictions)

same for the controlplane v1beta1 package

sbueringer commented on controlplane/kubeadm/internal/webhooks/kubeadm_control_plane.go:
```diff
@@ -70,26 +85,39 @@ func defaultKubeadmControlPlaneSpec(s *KubeadmControlPlaneSpec, namespace string
 
-func defaultRolloutStrategy(rolloutStrategy *RolloutStrategy) *RolloutStrategy {
+// DefaultKubeadmConfigSpec defaults a KubeadmConfigSpec.
+func DefaultKubeadmConfigSpec(r *bootstrapv1.KubeadmConfigSpec) {
```
Please move DefaultKubeadmConfigSpec from bootstrap/kubeadm/internal/webhooks to bootstrap/kubeadm/api/v1beta1 and then use it here.

sbueringer replied:
Same for the usage in KubeadmControlPlaneTemplate.Default

sbueringer commented on controlplane/kubeadm/internal/webhooks/kubeadmcontrolplanetemplate.go:
```diff
@@ -65,46 +82,52 @@ func (r *KubeadmControlPlaneTemplate) ValidateCreate() (admission.Warnings, erro
-	spec := r.Spec.Template.Spec
+	spec := k.Spec.Template.Spec
 	allErrs := validateKubeadmControlPlaneTemplateResourceSpec(spec, field.NewPath("spec", "template", "spec"))
 	allErrs = append(allErrs, validateClusterConfiguration(spec.KubeadmConfigSpec.ClusterConfiguration, nil, field.NewPath("spec", "template", "spec", "kubeadmConfigSpec", "clusterConfiguration"))...)
```
```suggestion
	allErrs = append(allErrs, validateClusterConfiguration(nil, spec.KubeadmConfigSpec.ClusterConfiguration,  field.NewPath("spec", "template", "spec", "kubeadmConfigSpec", "clusterConfiguration"))...)
```
Has to be adjusted as well as the order was flipped

sbueringer commented on controlplane/kubeadm/internal/webhooks/kubeadmcontrolplanetemplate.go:
```diff
@@ -65,46 +82,52 @@ func (r *KubeadmControlPlaneTemplate) ValidateCreate() (admission.Warnings, erro
+	if !reflect.DeepEqual(newK.Spec.Template.Spec, oldK.Spec.Template.Spec) {
 		allErrs = append(allErrs,
-			field.Invalid(field.NewPath("spec", "template", "spec"), r, kubeadmControlPlaneTemplateImmutableMsg),
+			field.Invalid(field.NewPath("spec", "template", "spec"), webhook, kubeadmControlPlaneTemplateImmutableMsg),
```
```suggestion
			field.Invalid(field.NewPath("spec", "template", "spec"), newK, kubeadmControlPlaneTemplateImmutableMsg),
```

sbueringer commented on controlplane/kubeadm/internal/filters.go:
```diff
@@ -215,15 +215,28 @@ func matchInitOrJoinConfiguration(machineConfig *bootstrapv1.KubeadmConfig, kcp
 }
 
+// defaultKubeadmConfigSpec defaults a KubeadmConfigSpec.
+func defaultKubeadmConfigSpec(r *bootstrapv1.KubeadmConfigSpec) {
```
Same here, we have to use  the same defaulting as in the KubeadmConfig webhook

sbueringer replied:
It is absolutely crucial that this code does not diverge 

If this runs out of sync KCP could trigger rollouts when none should be happening because it anticipates KubeadmConfigSpec defaulting incorrectly

sbueringer commented:
Thx! I think this goes definitely in the right direction. Should be another 1-2 rounds of reviews to find all the nits, but then we should be good

odvarkadaniel commented:
> Thx! I think this goes definitely in the right direction. Should be another 1-2 rounds of reviews to find all the nits, but then we should be good

Thank you for a review. I have addressed your comments.

JoelSpeed commented:
/lgtm

sbueringer commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfigtemplate.go:
```diff
@@ -14,52 +14,73 @@ See the License for the specific language governing permissions and
 }
 
-func (r *KubeadmConfigTemplateSpec) validate(name string) error {
+func (webhook *KubeadmConfigTemplate) validate(r bootstrapv1.KubeadmConfigTemplateSpec, name string) error {
```
```suggestion
func (webhook *KubeadmConfigTemplate) validate(r *bootstrapv1.KubeadmConfigTemplateSpec, name string) error {
```

sbueringer commented on controlplane/kubeadm/api/.import-restrictions:
```diff
@@ -0,0 +1,5 @@
+rules:
```
Let's also add a import-restrictions file to the v1beta1 package to block controller-runtime imports entirely there (like we did for the bootstrap provider v1beta1 package)

sbueringer reviewed (commented):
Last two nits :)

Feel free to squash directly

odvarkadaniel commented:
> /assign @killianmuldoon @chrischdi
> 
> Would be good if one of you can also take a look. I think this is the most complicated of the webhook refactorings, so want to make sure we don't introduce bugs

Addressed the 2 nits you mentioned above. Also squashed all the commits.

sbueringer commented:
Thank you very much!!!

/lgtm

chrischdi commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook_test.go:
```diff
@@ -337,56 +342,56 @@ func TestKubeadmConfigValidate(t *testing.T) {
 		},
 		"format is Ignition, experimental retry join is set": {
```
Can we readd this test?

sbueringer replied:
Are you looking at the latest state of the PR? This unit test got re-added already
(except if I"m missing something somehow :D)

odvarkadaniel replied:
This should already be re-added. https://github.com/kubernetes-sigs/cluster-api/blob/dd1e6bec00fbf45ea516ea79adb1a92b2fcbb479/bootstrap/kubeadm/internal/webhooks/kubeadmconfig_test.go#L343-L354

chrischdi replied:
Damn, local copy seems to be outdated :-(

chrischdi commented on controlplane/kubeadm/internal/webhooks/kubeadm_control_plane_test.go:
```diff
@@ -27,29 +27,36 @@ import (
+)
+
+var (
+	namespace = "bar"
```
Let's name it accordingly to what it is used for.
```suggestion
	invalidNamespaceName = "bar"
```

chrischdi commented:
Wow, that was quick

/lgtm

sbueringer commented:
@odvarkadaniel Thank you very much! That was one of the trickiest :)

Thx everyone for the reviews as well!

/lgtm
/approve
